File: java/formula-java/src/main/java/org/formula/parse/tree/RepeatingNode.java
package org.formula.parse.tree;

import java.util.List;
import java.util.Objects;

public class RepeatingNode<T> extends MatcherNode<T> {
    private final int minLength;
    private final int maxLength;

    public RepeatingNode(TokenMatcher matcher, int minLength, int maxLength) {
        super(matcher);
        this.minLength = minLength;
        this.maxLength = maxLength;
    }

    @Override
    public void walk(String text, int startIndex, int currentIndex, List<TokenMatch<T>> matches) {
        int minLength = this.minLength;
        int maxLength = calculateMaxLength(text, currentIndex);
        int length = 0;

        while (length < maxLength && currentIndex < text.length()) {
            char next = text.charAt(currentIndex);
            if (!matcher.matches(next)) {
                break;
            }
            currentIndex++;
            length++;
        }

        if (length < minLength) {
            return;
        }

        walkChildren(text, startIndex, currentIndex, matches);
        if (length > maxLength || mapper == null) {
            return;
        }

        TokenMatch<T> match = new TokenMatch<>(text, startIndex, currentIndex, mapper);
        matches.add(match);
    }

    @Override
    public boolean equals(Object other) {
        if (other == null) {
            return false;
        }
        if (getClass() != other.getClass()) {
            return false;
        }
        RepeatingNode<?> otherCasted = (RepeatingNode<?>) other;
        return Objects.equals(matcher, otherCasted.matcher)
                && Objects.equals(minLength, otherCasted.minLength)
                && Objects.equals(maxLength, otherCasted.maxLength);
    }

    private int calculateMaxLength(String text, int currentIndex) {
        int maxTextLength = text.length() - currentIndex;
        return Math.min(this.maxLength, maxTextLength);
    }
}

File: cs/formula-cs/Formula/TokenTree/RepeatingNode.cs
namespace Formula.TokenTree;

internal class RepeatingNode<T> : MatcherNode<T>
{
    private readonly int _minLength;
    private readonly int _maxLength;

    public RepeatingNode(ITokenMatcher matcher, int minLength, int maxLength) : base(matcher) {
        _minLength = minLength;
        _maxLength = maxLength;
    }
    
    public override void Walk(string text, int startIndex, int currentIndex, List<TokenMatch<T>> matches) {
        var minLength = _minLength;
        var maxLength = CalculateMaxLength(text, currentIndex);
        var length = 0;

        while (length < maxLength && currentIndex < text.Length) {
            char next = text[currentIndex];
            if (!Matcher.Matches(next)) 
            {
                break;
            }
            currentIndex++;
            length++;
        }

        if (length < minLength)
        {
            return;
        }

        WalkChildren(text, startIndex, currentIndex, matches);
        if (length > maxLength || Mapper == null) {
            return;
        }

        var match = new TokenMatch<T>(text, startIndex, currentIndex, Mapper);
        matches.Add(match);
    }
    
    public override bool Equals(object? other) {
        return other is RepeatingNode<T> node && Equals(node);
    }

    private bool Equals(RepeatingNode<T> other)
    {
        return base.Equals(other) && _minLength == other._minLength && _maxLength == other._maxLength;
    }

    public override int GetHashCode()
    {
        return HashCode.Combine(base.GetHashCode(), _minLength, _maxLength);
    }

    private int CalculateMaxLength(String text, int currentIndex) {
        int maxTextLength = text.Length - currentIndex;
        return Math.Min(_maxLength, maxTextLength);
    }
}